oonib/report/handlers.py

```diff
@@ -192,27 +140,63 @@ def parseNewReportRequest(request):
 
     return parsed_request
 
-class ReportHandler(OONIBHandler):
+def closeReport(report_id):
+    report_dir = FilePath(config.main.report_dir)
+    archive_dir = FilePath(config.main.archive_dir)
+    report_path = report_dir.child(report_id)
+    report_metadata_path = report_dir.child(report_id +
+                                            "-metadata.json")
+    print "Checking if %s exists and %s exists" % (report_path,
+                                                   report_metadata_path)
+    if not report_path.exists() or \
+            not report_metadata_path.exists():
+        raise e.ReportNotFound
+
+    with report_metadata_path.open('r') as f:
+        report_details = json.load(f)
+
+    dst_path = report_file_path(archive_dir,
+                                report_details,
+                                report_id)
+    if report_path.getsize() > 0:
+        report_path.moveTo(dst_path)
+    else:
+        # We currently just remove empty reports.
+        # XXX Maybe in the future we want to keep track of how often this
+        # happens.
+        report_path.remove()
+    report_metadata_path.remove()
+
+def checkForStaleReports():
+    report_dir = FilePath(config.main.report_dir)
+    for report_metadata_path in \
+            report_dir.globChildren("*-metadata.json"):
+        last_updated = time.time() - report_metadata_path.getModificationTime()
+        if last_updated > config.main.stale_time:
+            report_id = report_metadata_path.basename().replace(
+                '-metadata.json', '')
+            closeReport(report_id)
+    reactor.callLater(config.main.stale_time,
+                      checkForStaleReports)
```

Comment:
I suggest to schedule the callback as the first action of the callback such that, even in case of unexpected exception in the following code, we are robust in that the next scheduled callback will be called